src/Compiler/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/src/ViewComponentTagHelperDescriptorFactory.cs

```diff
@@ -58,30 +58,37 @@ public virtual TagHelperDescriptor CreateDescriptor(INamedTypeSymbol type)
         var tagName = $"vc:{HtmlConventions.ToHtmlCase(shortName)}";
         var typeName = $"__Generated__{shortName}ViewComponentTagHelper";
         var displayName = shortName + "ViewComponentTagHelper";
-        var descriptorBuilder = TagHelperDescriptorBuilder.Create(ViewComponentTagHelperConventions.Kind, typeName, assemblyName);
-        descriptorBuilder.SetTypeName(typeName);
-        descriptorBuilder.DisplayName = displayName;
-
-        if (TryFindInvokeMethod(type, out var method, out var diagnostic))
+        var descriptorBuilder = TagHelperDescriptorBuilder.GetInstance(ViewComponentTagHelperConventions.Kind, typeName, assemblyName);
+        try
```

Comment:
Not the end of the world, no. Do you have a suggested change? None of these methods (nor is this area) so high performance that removing try/finally blocks would have any noticeable impact.

Reply:
Note that the try/finally blocks make the code more maintainable. If someone were to add an optimization later to exit quickly with a return statement they might not notice that an object has been taken from a pool.

Reply:
Is your concern around the unwieldy-ness of the try/finally? Happy to add a disposable pattern and using statements if that's the case.

Reply:
> Note that the try/finally blocks make the code more maintainable. 

Agree. 

> Is it the end of the world if we don't return the builder in the exception case?

I know that in the Roslyn compiler we generally do not use `try / finally` for builders, pools, etc ... The rationale typically provided is roughly the following: 

1. This code can never throw an exception therefore `try / finally` is essentially instructions added to the code that should never be used. 
2. The Roslyn compiler was written for 10+ years without any `try / finally`. Introducing this pattern now is a massive change to the code base to catch all the places we do it. The code churn necessary to make it consistent is not worth the benefits it would provide. 

Of those arguments I've personally found (2) more persuasive. It would be a massive change to fully introduce it at this point and I have reservations about halfway introducing it. There are already parts of the compiler where we've half introduced patterns (aka testing) and it's bit us a number of times because copy / pasting code ends up changing it's execution.

I don't think the argument holds for Razor though. These pools are fairly new. Getting our code consistent isn't the massive undertaking it would be in the Roslyn compiler. I would personally focus on what is the better, more maintainable pattern and move forward with that. Not having to look at every new `return` in a PR and scan around to make sure the code didn't accidentally circumvent a pool being freed would be a nice change of pace compared to Roslyn :smile:

> Happy to add a disposable pattern and using statements if that's the case.

My 2 cents is I'd prefer a disposable pattern. Now that we have `using var` it means we get the `try / finally` behavior without the need for extra indentation in most cases. Also having an `IDisposable` return will push people to do the right thing in most cases as the type suggests a pattern.

Reply:
Implemented disposable pattern